Use hunk dependency code for computing inter-commit dependencies

mtsgrd opened the pull request:
Implementation for computing dependencies and inverse dependencies between commits.

@estib-vega:
The scope of this PR escalated a bit.
Through implementing this extra functionality we discovered a couple of bugs with the new dependency calculation algorithm.
This PR reimplements parts of it to account for different (less frequent but possible) use-cases and tests to ensure that that works.

### Changes
- Add integration tests for dependency calculation (inter commit and diffs to commit)
- Refactor the dependency calculation algo to account for use-cases like file deletion, recreation, lines added at the top, etc.

mtsgrd committed:
Use hunk dependency code for computing inter-commit dependencies

mtsgrd commented on crates/gitbutler-hunk-dependency/src/stack.rs:
```diff
@@ -47,4 +48,19 @@ impl StackRanges {
+            .iter()
+            .flat_map(|(_, path)| path.commits.iter())
+            .collect_vec();
+        data.into_iter()
+            .fold(HashMap::new(), |mut acc, (commit_id, dependencies)| {
+                acc.entry(*commit_id)
+                    .and_modify(|existing_dependencies| existing_dependencies.extend(dependencies))
+                    .or_insert(dependencies.clone());
+                acc
+            })
```
Just noticed, we declared the data variable for debugging, so we can drop that, skip the `.collect_vec()` and just return the concatenated the iter steps.

estib-vega committed:
Move out dependency calculation to a dedicated module

estib-vega committed:
Hunk locking: Remove the experimental flag

Move fully to the new Hunk Locking algorithm, remove the references to that flag and the settings UI

estib-vega committed:
Track dependencies in Stack

Stack keeps a map of commit and uncommitted chanegs inter-dependencies

estib-vega committed:
Propagate the dependency information to the commits

Propagate information about commit and uncommited change inter-dependencies to the commits for the FE to access

estib-vega commented:
@mtsgrd Since your changes I've done a couple of things:

### 1. We iterate over all files and commits when calculating the dependency graphs.
Before, we would only iterate over files that were present both in commits and in uncommitted changes because we were only interested on where the uncommitted files belonged.
Since we now are also interested on which commits depend on which, we need to traverse everything in the uncommitted files and stack commits.

### 2. I added a third hash-map that maps commit IDs to uncommitted hunks that depend on it
This enables to quickly check if a commit can't be moved because uncommitted hunks depend on it

### 3. Stack and Commits track dependency information 
Every time we call `get_applied_status_cached`, we'll calculate the dependency information and store it to each Stack respectively.
The Stack should be considered the source of truth.
This information is then propagated to the `VirtualBranchCommit` for the Front-End to access.

### Missing: Tests
I'll start adding some tests

estib-vega committed:
Dependencies: Rework of the algo

Re-implement the algorithm so that it takes account of different scenarios when comparing input diffs.
Those changes include:
- Awareness of change type 
- Detection of "special" diffs, like file deletion and file recreation
- Correctly calculate the commit dependency graph between commits

Also add and update unit and integration tests

estib-vega committed:
Dependencies: Add documentation

Add some high-level docs on the way the dependency graph is calculated

krlvi commented on crates/gitbutler-stack/src/stack.rs:
```diff
@@ -81,6 +82,17 @@ pub struct Stack {
     /// Do **NOT** edit this directly, instead use the `Stack` trait in gitbutler_stack.
     #[serde(default)]
     pub heads: Vec<Branch>,
+    // Dependency tracking
```
discussed IRL - lets not persist this

mtsgrd commented on crates/gitbutler-hunk-dependency/src/stack.rs:
```diff
@@ -23,13 +24,11 @@ impl StackRanges {
-            path_deps.add(stack_id, commit_id, diffs)?;
-            self.paths.insert(path.clone(), path_deps);
-        };
+        self.paths
+            .entry(path.to_owned())
+            .or_default()
+            .add(stack_id, commit_id, diffs)?;
```
🙌

krlvi commented on crates/gitbutler-serde/Cargo.toml:
```diff
@@ -10,3 +10,5 @@ git2.workspace = true
 gix.workspace = true
 serde = { workspace = true, features = ["std"]}
 bstr.workspace = true
+md5 = "0.7.0"
```
likely this is no longer needed

krlvi commented on crates/gitbutler-branch-actions/src/commit.rs:
```diff
@@ -50,8 +51,19 @@ pub struct VirtualBranchCommit {
+    pub commit_dependencies: Vec<git2::Oid>,
+    /// Commits that depend on this commit.
+    #[serde(default, with = "gitbutler_serde::oid_vec")]
+    pub inverse_commit_dependencies: Vec<git2::Oid>,
```
Often in the context of dependencies, the word "reverse dependencies" is used

estib-vega committed:
Dependencies: Update the algo terminology

estib-vega committed:
Dependencies: Algo tracks indices and shifts correctly

Track what's the index of:
- The next hunk range to visit
- The index of the the first hunk range to shift after adding an incoming hunk.

Also, calculate correctly the lines of the hunk ranges that are trimmed at the top in the following cases:
- Incomming hunk is only deleting lines
- Incomming hunk is only adding lines

Add special handling for intersections with hunks that only add lines

estib-vega reviewed (approved):
I'm biased, but looks good to me :P

estib-vega commented on crates/gitbutler-hunk-dependency/README.md:
```diff
@@ -0,0 +1,213 @@
+# Hunk dependency
```
Oh, that's actually intended.
That's how you can create multi-line paragraphs.

estib-vega committed:
Address feedback

mtsgrd commented on apps/desktop/src/lib/vbranches/types.ts:
```diff
@@ -239,15 +239,15 @@ export class DetailedCommit {
 	 * The ids of the commits that depend on this commit.
 	 */
-	inverseCommitDependencies!: string[];
+	inverseDependencies!: string[];
```
As @krlvi suggested, should we perhaps call this `reverseDependencies`?

estib-vega replied:
Right, I'll change it to that 👍

mtsgrd commented on crates/gitbutler-branch-actions/src/dependencies.rs:
```diff
@@ -24,7 +28,7 @@ pub fn compute_workspace_dependencies(
     for stack in stacks {
         let mut commits_input: Vec<InputCommit> = vec![];
-        // Commit IDs in application order
+        // The commit id's must be ordered from order they get applied.
```
It feels a bit like there's something funny in this sentence, would it be clearer with this? `// Commit id's must be ordered tail first.`

estib-vega replied:
Hmm, tail introduces more questions imo.
How about talking about parentage?
`// Commit id's must be ordered ancestors-to-descendants`
or 
`// Commit id's are expected to be ordered parents first.`?

mtsgrd commented on crates/gitbutler-branch-actions/src/status.rs:
```diff
@@ -98,18 +92,10 @@ pub fn get_applied_status_cached(
-        let base_tree = ctx.repository().find_commit(default_target.sha)?.tree()?;
-        compute_old_locks(ctx.repository(), &base_diffs, &virtual_branches, base_tree)?
-    };
+    let workspace_dependencies =
+        compute_workspace_dependencies(ctx, &default_target.sha, &base_diffs, &virtual_branches)?;
```
@krlvi  I think we discussed a thing related to this line earlier today, would you prefer ok or default here as to not give the user a "something went wrong"?

mtsgrd commented:
Starting to look pretty good to me. Let's get it into the next nightly?
